types/errors/proto.go

```diff
@@ -0,0 +1,56 @@
+package errors
+
+import (
+	"fmt"
+	"reflect"
+)
+
+type ErrMsgToProto struct {
+	MessageName string
+	Err         error
+}
+
+func NewErrMsgToProto[T any](msg T, err error) ErrMsgToProto {
```

Comment:
To reviewers:

what is the opinion on this generic factory? This way there is no need to provide a string with struct name when returning the error, you can just pass the struct that caused the error.

Reply:
if this design is desired, I could change [other](https://github.com/cometbft/cometbft/pull/1211/files#diff-f8a280b25c2ff3601ed557585a8ba277859b7782cdf5a277931dc62721462515R17) errors to use the same pattern

Reply:
Given that all we're really using the `msg` field for is its name, I'd say the use of generics and reflection here is undesirable. Especially given that one has to instantiate the empty struct as well when calling the constructor in some cases.

My take is that supplying the type name as a string is simpler, and therefore better. Of course, if we had to rename any of the structs, we'd need to remember to rename the strings, but the chances of us needing to rename such structs often is _very_ low.

If we were embedding the source message itself in the error, then this approach could possibly make sense. But it's not clear whether that's desirable.

Reply:
Thanks for a clear feedback, I will refactor it to the desirable design!